gubernator/github_auth.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2016 The Kubernetes Authors All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Copyright 2016 The Kubernetes Authors All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import urllib
+
+from webapp2_extras import security
+
+from google.appengine.api import urlfetch
+
+import view_base
+
+
+class Endpoint(view_base.BaseHandler):
+    def github_client(self):
+        client = self.app.config['github_client']
+        if client is None:
+            self.abort(500)
+        return client['id'], client['secret']
+
+    def maybe_redirect(self, target):
+        '''
+        Redirect to a given URL if it's determined to be safe.
+        '''
+        if target.startswith('/pr'):
+            self.redirect(target)
+
+    def get(self, arg):
+        # Documentation here: https://developer.github.com/v3/oauth/
+
+        client_id, client_secret = self.github_client()
+
+        if arg.endswith('/done'):
+            target, done = arg[:-len('/done')], True
+        else:
+            target, done = arg, False
+
+        if not done:
+            # 1) Redirect users to request GitHub access
+            if self.session.get('user'):
+                # User already logged in, no need to continue.
+                self.maybe_redirect(target)
+                return
+
+            state = security.generate_random_string(entropy=128)
+            args = {
+                'client_id': client_id,
+                'redirect_uri': self.request.url + '/done',
+                'scope': '',  # Username only needs "public data" permission!
+                'state': state
+            }
+            self.session['gh_state'] = state
+            self.redirect('https://github.com/login/oauth/authorize?'
```

Comment:
Hmm... I recommend the habit of just always using % rather than switching back and forth between + and % depending on whether you're combining 2 or more than two strings.

PS: If you keep it like this will you fix the indentation? (Line break before the start of the string)